In [84]:
import pandas as pd 
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

# 1.

In [51]:
df = pd.read_csv('./Prestige.csv')
df.head()

Unnamed: 0  education  income  women  prestige  census  type
0   gov.administrators      13.11   12351  11.16      68.8    1113  prof
1     general.managers      12.26   25879   4.02      69.1    1130  prof
2          accountants      12.77    9271  15.70      63.4    1171  prof
3  purchasing.officers      11.42    8865   9.11      56.8    1175  prof
4             chemists      14.62    8403  11.68      73.5    2111  prof

In [52]:
df = df.drop(columns=['Unnamed: 0','type'])
df.head()

education  income  women  prestige  census
0      13.11   12351  11.16      68.8    1113
1      12.26   25879   4.02      69.1    1130
2      12.77    9271  15.70      63.4    1171
3      11.42    8865   9.11      56.8    1175
4      14.62    8403  11.68      73.5    2111

In [59]:
Y = df['prestige'].to_numpy()
X = df.drop(columns=['prestige']).to_numpy()
model0 = LinearRegression()
model0.fit(X, Y)
model0.coef_

array([ 4.65715805e+00,  1.28922425e-03, -2.08682032e-03,  5.68420955e-04])

$\beta_{0,1}=4.65715805$

i. Realiza la regresión de $Y$ usando las variables $X_2 , ..., X_m$ , obteniendo los residuos $E_{Y|X_2 ,...,X_m}$ .

In [53]:
Y = df['prestige'].to_numpy()
X = df.drop(columns=['education', 'prestige']).to_numpy()
model1 = LinearRegression()
model1.fit(X, Y)
model1.coef_

array([ 0.00237824,  0.02549987, -0.00267957])

In [54]:
Y_pred = model1.predict(X)
EYX2Xm = (Y - Y_pred).reshape(-1, 1)

ii. Realiza la regresión de $X_1$ usando las variables $X_2, ..., X_m$ , obteniendo los residuos $E_{X_1|X_2,...,X_m}$.

In [ ]:
X1 = df['education'].to_numpy()
X = df.drop(columns=['prestige', 'education']).to_numpy()
model2 = LinearRegression()
model2.fit(X, X1)
model2.coef_

In [56]:
X1_pred = model2.predict(X)
EX1X2Xm = (X1 - X1_pred).reshape(-1, 1)

iii. Realiza la regresión de la variable formada por los residuos $E_{Y|X_2,...,X_m}$ usando la variable $E_{X_1|X_2,...,X_m}$. La pendiente de esta regresión es exactamente $β_{0,1}$.

In [57]:
model3 = LinearRegression()
model3.fit(EX1X2Xm, EYX2Xm)
model3.coef_

array([[4.65715805]])

$pendiente=4.65715805$

(a) $\beta_{0,1}=pendiente$

In [60]:
model3.intercept_

array([-3.50520587e-15])

$intercepto=-3.50520587\times 10^{-15}$

(b) claramente $intercepto \approx 0$, en clase vimos que para cualquier regresión lineal, $\overline{E} = \frac{1}{N} \sum_{i=1}^N y_i - \hat{y}_i = 0$ y sabemos que

$intercepto = \overline{EYX2Xm} - pendiente*\overline{EX1X2Xm}$


# 2.

In [73]:
np.corrcoef(EX1X2Xm.flatten(), EYX2Xm.flatten())

array([[1.        , 0.62048791],
       [0.62048791, 1.        ]])

(a) $corr=0.62048791$

(b) $(\Rightarrow)$ Si $\beta_{0,1}=0$, por el el inciso (b) del ejercicio anterior $pendiente=0$, entonces $0=\frac{\sum_{i=1}^N (E_{i, X_1|X_2,\dots, X_m} - \overline{E}_{X_1|X_2,\dots, X_m})(E_{i, Y|X_2,\dots, X_m} - \overline{E}_{Y|X_2,\dots, X_m})}{\sum_{i=1}^N (E_{i, X_1|X_2,\dots, X_m} - \overline{E}_{X_1|X_2,\dots, X_m})^2}$. Por lo tanto, $\sum_{i=1}^N (E_{i, X_1|X_2,\dots, X_m} - \overline{E}_{X_1|X_2,\dots, X_m})(E_{i, Y|X_2,\dots, X_m} - \overline{E}_{Y|X_2,\dots, X_m})=0$ y así $corr=\frac{\sum_{i=1}^N(E_{i, X_1|X_2,\dots, X_m} - \overline{E}_{X_1|X_2,\dots, X_m})(E_{i, Y|X_2,\dots, X_m} - \overline{E}_{Y|X_2,\dots, X_m})}{\sqrt{\sum_{i=1}^N(E_{i, X_1|X_2,\dots, X_m} - \overline{E}_{X_1|X_2,\dots, X_m})^2}\sqrt{\sum_{i=1}^N(E_{i, Y|X_2,\dots, X_m} - \overline{E}_{Y|X_2,\dots, X_m})^2}}=0$.

El regreso es análogo.

# 3.

In [76]:
df = pd.read_csv('./50_Startups.csv')
df.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [78]:
df = df.drop(columns=['State'])
df.head()

R&D Spend  Administration  Marketing Spend     Profit
0  165349.20       136897.80        471784.10  192261.83
1  162597.70       151377.59        443898.53  191792.06
2  153441.51       101145.55        407934.54  191050.39
3  144372.41       118671.85        383199.62  182901.99
4  142107.34        91391.77        366168.42  166187.94

(a) Las variables independientes son *R&D Spend*, *Administration*, *Marketing Spend* y la variable dependiente es *Profit*

In [79]:
df.corr()

R&D Spend  Administration  Marketing Spend    Profit
R&D Spend         1.000000        0.241955         0.724248  0.972900
Administration    0.241955        1.000000        -0.032154  0.200717
Marketing Spend   0.724248       -0.032154         1.000000  0.747766
Profit            0.972900        0.200717         0.747766  1.000000

(b) La correlación entre *Marketing Spend* y *R&D Spend* podría considerarse alta, por lo que podríamos optar por eliminar una de ellas. Claramente *R&D Spend* está mas correlacionada con la variable dependiente que *Marketing Spend*

In [80]:
df = df.drop(columns=['Marketing Spend'])

In [92]:
df.corr()

R&D Spend  Administration    Profit
R&D Spend        1.000000        0.241955  0.972900
Administration   0.241955        1.000000  0.200717
Profit           0.972900        0.200717  1.000000

(c)

In [83]:
X = df.drop(columns=['Profit'])
Y = df['Profit']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=0)

In [90]:
model = LinearRegression()
model.fit(X_train, Y_train)
Y_pred = model.predict(X_test)
r2 = r2_score(Y_test, Y_pred)
print(r2)

0.9469407189577183


(d) $R^2=0.9469407189577183$ 